File: tests/unit/platform/linux/test_profile_sources.c
#define main profile_import_main
#include "packaging/linux/vibepollo-profile-import.c"
#undef main
#include <assert.h>

int main(void) {
  char fixture[] = "/tmp/vibepollo-profile-sources.XXXXXX";
  assert(mkdtemp(fixture));
  int home = open(fixture, O_RDONLY | O_DIRECTORY | O_CLOEXEC);
  assert(home >= 0);
  assert(mkdirat(home, ".config", 0700) == 0);
  assert(open_source_profile(home, "auto") == -1 && errno == ENOENT);
  assert(mkdirat(home, ".config/vibeshine", 0700) == 0);
  int source = open_source_profile(home, "auto");
  assert(source >= 0); close(source);
  assert(mkdirat(home, ".config/sunshine", 0700) == 0);
  assert(open_source_profile(home, "auto") == -1 && errno == EEXIST);
  source = open_source_profile(home, "sunshine");
  assert(source >= 0); close(source);
  assert(unlinkat(home, ".config/sunshine", AT_REMOVEDIR) == 0);
  assert(symlinkat("/tmp", home, ".config/sunshine") == 0);
  assert(open_source_profile(home, "auto") == -1 && (errno == ELOOP || errno == ENOTDIR));
  assert(unlinkat(home, ".config/sunshine", 0) == 0);
  assert(unlinkat(home, ".config/vibeshine", AT_REMOVEDIR) == 0);
  assert(unlinkat(home, ".config", AT_REMOVEDIR) == 0);
  close(home);
  assert(rmdir(fixture) == 0);
  puts("Legacy profile selection, ambiguity and symlink refusal passed.");
}

File: tests/unit/platform/linux/test_profile_sources.py
#!/usr/bin/env python3
"""Exercise confined legacy source discovery as the ordinary test user."""
import json
from pathlib import Path
import subprocess
import tempfile

ROOT = Path(__file__).resolve().parents[4]
with tempfile.TemporaryDirectory() as temporary:
    base = Path(temporary)
    harness = base / 'test.c'
    harness.write_text('#define main importer_main\n#include ' +
                       json.dumps(str(ROOT / 'packaging/linux/vibepollo-profile-import.c')) +
                       '\n#undef main\n' + r'''
int main(int argc, char **argv) {
  int home = open(argv[1], O_RDONLY | O_DIRECTORY);
  int source = open_source_profile(home, "auto");
  if (source >= 0) { close(source); close(home); return 0; }
  int error = errno;
  close(home);
  return error == ENOENT ? 10 : error == EEXIST ? 11 : 12;
}
''')
    subprocess.run(['cc', '-Wall', '-Wextra', '-Wno-unused-parameter', str(harness),
                    '-lcap', '-o', str(base / 'test')], check=True)
    home = base / 'home'
    home.mkdir()
    (home / '.config').mkdir()
    def check(expected):
        assert subprocess.run([str(base / 'test'), str(home)]).returncode == expected
    check(10)
    (home / '.config/sunshine').mkdir()
    check(0)
    (home / '.config/vibeshine').mkdir()
    check(11)
    (home / '.config/sunshine').rmdir()
    check(0)
    (home / '.config/sunshine').symlink_to('vibeshine', target_is_directory=True)
    check(12)
print('PASS: fresh, Sunshine, Vibeshine, ambiguous and symlink profile sources')
